Block brake improvements

spacek531 opened the pull request:
I started on this branch because at the time I didn't know about the old PR. @X123M3-256 told me about the old PRs https://github.com/OpenRCT2/OpenRCT2/pull/7737 and https://github.com/OpenRCT2/OpenRCT2/issues/11649 and that there were some larger technical issues, but he didn't remember what they were and the old PR does not say what they are.

OpenRCT2 side, everything is unified to the block brake flag indicating that brakes are active and the absence meaning they are open. 
SV6 side, the polarity is inverted for brakes, since they normally serialize with the flag as false.
TD6 side, block brake speed is not exported nor imported. It is set to the minimum on loading.
NSF side, a version bump is required and behavior to normalize old save format brake flags.

spacek531 commented:
[ ] reset brake open status on removing vehicles or entering construction mode

spacek531 commented:
TD6 are not capable of saving both rotation and speed in the same piece. If this limitation matters, disabling 4D rotation on speedy block brakes is the better course of action until new track design format is implemented.

spacek531 commented:
Additionally, TD6 write the speed to block brakes. I didn't look at the TD6 format to see if there's one bit free, so for now all block brakes import from TD6 at the default behavior regardless of what they were saved as.

spacek531 commented:
Considering re-implementing or rebasing this onto new-save-format. Marking as draft for now.

spacek531 commented:
For some reason the related issues isn't showing up in the sidebar. Here's hoping this fixes it:

https://github.com/OpenRCT2/OpenRCT2/issues/8570

https://github.com/OpenRCT2/OpenRCT2/issues/11649

https://github.com/OpenRCT2/OpenRCT2/pull/7737

spacek531 commented:
Maintaining this history has become untenable in light of the random seed being affected. I am going to start from develop and add changes in a way that is hopefully easier to follow and debug, then force-push onto this branch.

spacek531 commented:
Something has happened to OpenRCT2/develop to make this desync hard and has numerous graphical glitches. When re-implementing this branch one change at a time after the rebase, the change to `TrackTypeHasSpeedSetting()` in `Track.cpp` caused the srand seed to change but something else caused the graphical glitches.

spacek531 commented:
With some testing I discovered that the minimum amount of code required to create this big desync behavior is just 5 lines: a7a0441c338240b7d1e5b56985fce013ecd319e4

By checking out each commit on the develop branch one-by-one and cherry-picking this commit onto it, I discovered the first commit to cause the seed change is 78f6e3e8e36800d016ca7694c40320349acddca5.

There is no possible way this feature can be implemented without the 5 lines changing as seen in the aforementioned commit. The question is where to proceed from here?

spacek531 commented:
I tested to see if checking out the latest commit in develop 78f6e3e8e36800d016ca7694c40320349acddca5, reverting78f6e3e8e36800d016ca7694c40320349acddca5 , then cherry-picking a7a0441 would give me the expected desync file of 2kb. I solved both merge conflicts in favor of HEAD.

It generated a desync file of 3kb, and srand matched. 3kb is a bit larger than 2kb, but nowhere close to 8000kb. As expected, the only changes in the desync file were of vehicles' `brake_speed` being set to `0x02` when previously `0x00`.

Something specific about that commit has changed srand significantly.

spacek531 commented:
The problem with the desyncs has been resolved. Back to creating manageable desyncs of 2-3kb in size.

spacek531 commented:
Changing the tile inspector text based on what track piece is selected:

![openrct2_2021-01-22_17-04-44](https://user-images.githubusercontent.com/3454156/105563760-26e07b00-5cd4-11eb-89db-43819996322c.png)
![openrct2_2021-01-22_17-04-51](https://user-images.githubusercontent.com/3454156/105563784-32cc3d00-5cd4-11eb-80ea-aa3d95395e4a.png)
![openrct2_2021-01-22_17-04-31](https://user-images.githubusercontent.com/3454156/105563792-4081c280-5cd4-11eb-8394-3456e6f8518e.png)

spacek531 commented:
I have rebased this PR onto #13857; when it is merged, I will force-push this branch to the latest version.

Gymnasiast commented on src/openrct2/ride/Track.cpp:
```diff
@@ -1366,20 +1366,20 @@ void TrackElement::SetInverted(bool inverted)
 }
 
-bool TrackElement::BlockBrakeClosed() const
+bool TrackElement::GetBrakeClosed() const
```
IsBrakeClosed()

spacek531 commented:
If it's fine to export brakes with inverted `isClosed` polarity, then this belongs in Develop. If save file versioning is required, then this belongs in NSF.

spacek531 commented on src/openrct2/ride/TrackDesign.cpp:
```diff
@@ -200,15 +200,18 @@ rct_string_id TrackDesign::CreateTrackDesignTrack(const Ride& ride)
 
         uint8_t trackFlags;
-        if (TrackTypeHasSpeedSetting(track.type))
+#ifdef EXPORT_AS_TD6
```
It would be better if TrackDesign structs were improved for a NTDF and make T6Import/Export do more work.

tupaschoal commented:
Is this ready for review?

spacek531 commented:
There's one or two vestigial pieces that need to be removed, but I think it's ready.

spacek531 commented on src/openrct2/rct2/S6Importer.cpp:
```diff
@@ -1272,7 +1272,7 @@ namespace RCT2
                     dst2->SetStationIndex(src2->GetStationIndex());
                     dst2->SetHasGreenLight(src2->HasGreenLight());
-                    dst2->SetBlockBrakeClosed(src2->BlockBrakeClosed());
+                    dst2->SetBrakeClosed(src2->GetBrakeClosed() || (trackType == TrackElemType::Brakes));
```
TODO: remove this because it's taken care of in GetBrakeClosed

spacek531 replied:
It is not taken care of in GetBrakeClosed

spacek531 commented on src/openrct2/ride/Ride.cpp:
```diff
@@ -3565,7 +3659,10 @@ void Ride::MoveTrainsToBlockBrakes(TrackElement* firstBlock)
 
-        firstBlock->SetBlockBrakeClosed(true);
+        firstBlock->SetBrakeClosed(true);
+        blockBrakeSetLinkedBrakesClosed(
```
Could add a check here to determine if it's a TrackElemType::BlockBrake

spacek531 replied:
It is always a block brake because VEHICLE_UPDATE_MOTION_TRACK_FLAG_VEHICLE_AT_BLOCK_BRAKE is only tripped when it's a block brake

rik-smeets commented:
I'm closing this PR due to inactivity, feel free to reopen when you continue working on it.

spacek531 commented:
I have rebased and addressed the outstanding issues. I would reopen this PR but since I am not a maintainer, I am not able to. Would someone please re-open this PR?

tupaschoal commented:
I can't:
![image](https://user-images.githubusercontent.com/6443427/204267202-7900b2b9-0085-4e41-aa2e-e6bf07762d37.png)
